DESIGN.md

```diff
@@ -0,0 +1,155 @@
+# MatchTree Design Documentation
+
+This document is to serve as an external explanation of the design, model, and API of the
+`MatchTree` infrastructure within `doodle`. It should be viewed as a living document, and may not
+precisely reflect the most up-to-date implementation as details change. However, it is easier to include
+longer digressions on the model of a given type or operation within the system, in a separate document,
+rather than crowd the definition sites in the library source file itself.
+
+## Terminology
+
+Abstractly, we may refer to the model of a `MatchTree` in terms of a trie whose nodes are *choice points* and whose children are *conditioned* on a particular set of values for the immediate byte of input at the current lookahead-offset into the buffer being processed.
+
+By 'choice point', we mean an individual multi-way decision between one of K possible one-step descents.
+Each such descent is associated (and conditioned on) a particular set of byte-values, which are mutually disjoint across any two branches of the same choice-point.
+
+We may refer to one or more of the byte-sets upon which a choice-point is conditioned, as the *branching-predicates* of said node.
+
+## Core Types
+
+The `MatchTree` design consists of three core types, and one auxiliary type:
+
+* [`MatchTree`](#matchtree-type-definition): a finalized prefix-tree evaluated to an abstract, but consistent depth across all branches (modulo short-circuit pruning in cases where acceptance or rejection is unconditional)
```

Comment:
What exactly makes a depth "abstract but consistent"?

Reply:
Perhaps there is a better way of phrasing it, but what I mean is, every branch $B_i$ is evaluated to a depth $N_i$, where $\exists N^\star \ni \forall i. N = N_i $ but the actual value of $N$ is not available, and may be unknowable in some cases (e.g. if all branches accept within $M < N$ bytes of input). If you can come up with a more direct way of getting that idea across, or if you think it is not necessary to mention, I would be happy to make the appropriate changes.